cmd/codegen/generator/go/templates/module_funcs.go

```diff
@@ -326,6 +336,10 @@ type paramSpec struct {
 
 	optional bool
 	variadic bool
+	// hasOptionalWrapper is true if the type is wrapped in the Optional generic type
+	hasOptionalWrapper bool
```

Comment:
Hm, is this required? The idea is that `paramSpec` should have the raw type that was actually read from the argument, so we don't need to track this separately.

Maybe it is necessary for use in `concreteMethodCode`, but I'm not 100% sure why this case is used, and wasn't before.

I guess the same logic applies to `isPtr`.

Reply:
> Hm, is this required? The idea is that paramSpec should have the raw type that was actually read from the argument, so we don't need to track this separately.
Maybe it is necessary for use in concreteMethodCode, but I'm not 100% sure why this case is used, and wasn't before.

It's necessary now because we need to actually generate code that wraps a type in `Optional` whereas before all we cared about was marking something as optional in the `TypeDef` (which is just a bool, so it doesn't matter there whether it's optional due to being wrapped in `Optional` or due to an `//+optional` pragma).

The other possibility was to not add `hasOptionalWrapper` and instead just use `paramType` and re-run the logic of checking whether that's `Optional`. But it just felt simpler to have an extra field here since we already did that checking in `parseParamSpecVar`, may as well just re-use the work done there rather than re-run the logic a bunch of places (and handle `if err != nil `, etc.)